storage/lmdb/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod error;
+
+use std::path::PathBuf;
+
+use lmdb::Cursor;
+use storage_core::{
+    backend::{self, TransactionalRo, TransactionalRw},
+    info::{DbDesc, MapDesc},
+    Data, DbIndex,
+};
+use utils::sync::Arc;
+
+/// Metadata about the list of databases (key-value maps)
+#[derive(Eq, PartialEq, Debug, Clone)]
+struct DbList(Vec<lmdb::Database>);
+
+impl std::ops::Index<DbIndex> for DbList {
+    type Output = lmdb::Database;
+
+    fn index(&self, index: DbIndex) -> &Self::Output {
+        &self.0[index.get()]
+    }
+}
+
+pub struct PrefixIter<'tx> {
+    iter: lmdb::Iter<'tx>,
+    prefix: Data,
+    // Cursor needs to be kept alive for the duration of the iterator lifetime
+    _cursor: lmdb::RoCursor<'tx>,
+}
+
+impl<'tx> PrefixIter<'tx> {
+    fn new(cursor: lmdb::RoCursor<'tx>, iter: lmdb::Iter<'tx>, prefix: Data) -> Self {
+        PrefixIter {
+            _cursor: cursor,
+            iter,
+            prefix,
+        }
+    }
+}
+
+impl Iterator for PrefixIter<'_> {
+    type Item = (Data, Data);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (k, v) = self.iter.next()?.expect("iteration to proceed");
+        utils::ensure!(k.starts_with(&self.prefix));
+        Some((k.to_vec(), v.to_vec()))
+    }
+}
+
+pub struct DbTx<'m, Tx> {
+    tx: Tx,
+    dbs: &'m DbList,
+}
+
+type DbTxRo<'a> = DbTx<'a, lmdb::RoTransaction<'a>>;
+type DbTxRw<'a> = DbTx<'a, lmdb::RwTransaction<'a>>;
+
+impl<'s, 'i, Tx: lmdb::Transaction> backend::PrefixIter<'i> for DbTx<'s, Tx> {
+    type Iterator = PrefixIter<'i>;
+
+    fn prefix_iter<'t: 'i>(
+        &'t self,
+        idx: DbIndex,
+        prefix: Data,
+    ) -> storage_core::Result<Self::Iterator> {
+        let mut cursor = self.tx.open_ro_cursor(self.dbs[idx]).or_else(error::process_with_err)?;
+        let iter = if prefix.is_empty() {
+            cursor.iter_start()
```

Comment:
I would not make that distinction... the iter_from should be good enough for all cases.

Reply:
The homogeneity of this system is very important. If it turns out there's different treatment for different calls, then it's gonna be hard to figure out what's going on at the user level.

Reply:
> I would not make that distinction... the iter_from should be good enough for all cases.

That's what I had initially but it did not work. LMDB does not support empty keys, so it has to be done this way as far as I can tell.

Reply:
Alrighty.